File: frontend/public/js/secant.js
import { javaUrl, rustUrl } from '/config.js'

document.addEventListener("DOMContentLoaded", function() {

    if (!window.location.pathname.startsWith('/secant')) {
        return;
    }

    console.log("Secant Method page loaded");

    const form = document.getElementById('secant-form');
    const functionInput = document.getElementById('function-input');
    const latexOutput = document.getElementById('latex-output');

    var functionValue;

    form.addEventListener('submit', function(event) {
        event.preventDefault();
        
        const formData = new FormData(form);
        functionValue = formData.get('function');
        const params = new URLSearchParams({
            function: formData.get('function'),
            initialGuess1: formData.get('initialGuess1'),
            initialGuess2: formData.get('initialGuess2'),
            precisionType: formData.get('precisionType'),
            errorType: formData.get('errorType'),
            toleranceValue: formData.get('toleranceValue'),
            maxIterations: formData.get('maxIterations')
        });

        fetch(`${javaUrl}/api/v1/secant?${params.toString()}`, {
            method: 'POST'
        })
        .then(response => response.json())
        .then(result => {
            const prettyResult = JSON.stringify(result, null, 2);
            console.log('Formatted Result:', prettyResult);
            displayResult(result);
        })
        .catch(error => {
            console.error('Error:', error);
        });
    });

    function displayResult(result) {
        const resultText = result.message;
        document.getElementById('result-text').innerHTML = resultText;
        
        const resultTable = document.getElementById('result-table');
        resultTable.innerHTML = ''; // Limpiar tabla

        if (result.iterations && result.xvalues && result.functionValues && result.errors) {
            for (let i = 0; i < result.iterations.length; i++) {
                const row = document.createElement('tr');
                row.innerHTML = `
                    <td>${i}</td>
                    <td>${result.xvalues[i]}</td>
                    <td>${result.functionValues[i]}</td>
                    <td>${result.errors[i]}</td>
                `;
                resultTable.appendChild(row);
            }

            const s = result.xvalues[result.xvalues.length - 1];
            console.log("S=" + s);
            plotFunction(functionValue, s);
        } else {
            console.error('Invalid result data', result);
            alert('Invalid result data');
        }
    }

    functionInput.addEventListener('input', function() {
        const latexString = toLaTeX(functionInput.value);
        latexOutput.textContent = `$$${latexString}$$`;
        MathJax.Hub.Queue(["Typeset", MathJax.Hub, latexOutput]);
    });

    function plotFunction(mathFunction, s) {
        const xvals = [];
        const yvals = [];
        const scope = {};

        console.log("Mathfunction=" + mathFunction);
        const node = math.parse(mathFunction);

        const startX = s - 20;
        const endX = s + 20;
        const step = 0.1;

        for (let x = startX; x <= endX; x += step) {
            scope.x = x;
            try {
                const y = node.evaluate(scope);
                xvals.push(x);
                yvals.push(y);
            } catch (err) {
                console.error(`Error evaluating function at x=${x}:`, err);
            }
        }

        const trace = {
            x: xvals,
            y: yvals,
            type: 'scatter'
        };

        const layout = {
            title: 'Function Graphic',
            xaxis: {
                title: 'x'
            },
            yaxis: {
                title: 'f(x)'
            }
        };

        Plotly.newPlot('result-chart', [trace], layout);
    }

    function toLaTeX(mathExpr) {
        try {
            const node = math.parse(mathExpr);
            return node.toTex();
        } catch (err) {
            console.error('Error parsing math expression:', err);
            return mathExpr;
        }
    }
});

File: backend/python/methods/secant.py
import sympy as sp
import math

def secant(function_expr, initial_guess1, initial_guess2, error_type, tolerance_value, max_iterations):
    x = sp.symbols('x')
    function = sp.sympify(function_expr)
    
    tolerance = 0.5 * 10 ** -tolerance_value
    
    xvalues = []
    functionvalues = []
    errors = []
    iterations = []

    x0 = initial_guess1
    x1 = initial_guess2
    f0 = function.evalf(subs={x: x0})
    f1 = function.evalf(subs={x: x1})
    iteration_count = 0
    error = 100.0  # Initial error set to 100%
    xvalues.append(x0)
    xvalues.append(x1)
    functionvalues.append(f0)
    functionvalues.append(f1)
    errors.append(error)
    errors.append(abs(x1 - x0))
    iterations.append(0)
    iterations.append(1)

    if math.isnan(f0) or math.isnan(f1) or math.isinf(f0) or math.isinf(f1):
        message = f"The function is not defined at x0 = {initial_guess1} or x1 = {initial_guess2}. The method fails."
        return {
            "message": message,
            "xvalues": [],
            "functionValues": [],
            "errors": [],
            "iterations": []
        }

    while iteration_count < max_iterations and errors[-1] >= tolerance and f1 != 0 and f1 - f0 != 0:
        iteration_count += 1
        x2 = x1 - (f1 * (x1 - x0)) / (f1 - f0)
        x0 = x1
        f0 = f1
        x1 = x2
        f1 = function.evalf(subs={x: x1})

        if math.isnan(f1) or math.isinf(f1):
            message = f"The function is not defined at x = {x1}. The method fails."
            return {
                "message": message,
                "xvalues": xvalues,
                "functionValues": functionvalues,
                "errors": errors,
                "iterations": iterations
            }

        xvalues.append(x1)
        functionvalues.append(f1)
        iterations.append(iteration_count)

        error = abs(x1 - x0) if error_type == 1 else abs((x1 - x0) / x1)
        errors.append(error)

    message = f"{x1} is a root of f(x)" if f1 == 0 else f"The approximate solution is: {x1}, with a tolerance = {tolerance}" if errors[-1] < tolerance else f"Failed in {max_iterations} iterations"
    return {
        "message": message,
        "xvalues": xvalues,
        "functionValues": functionvalues,
        "errors": errors,
        "iterations": iterations
    }